### PR TITLE
Update apis 

### PR DESCRIPTION
**Pull request by sandrodevdariani**

### CODE REVIEW

File: orchestration/api/api_classifier_score.py [hunk 1]
```diff
@@ -686,6 +714,7 @@ async def set_image_classifier_score_list(request: Request, classifier_score_lis
                     http_status_code=404
                 )
             image_hash = job_data['task_output_file_dict']['output_file_hash']
+            image_uuid = job_data.get('image_uuid', None)
```
Comment (Senyoret1) [inline orchestration/api/api_classifier_score.py:717]: I think this won't work, because the query is not retrieving the image_uuid from the database. This would mean saving "null" in the database.

File: orchestration/api/api_external_images.py [hunk 2]
```diff
@@ -343,25 +342,48 @@ async def get_external_image_list_without_extracts(request: Request, dataset: st
...
 
-        # Apply sampling if the size parameter is provided
+        image_data_list = []
+        batch_size = 1000  # Set the batch size
+        accumulated_size = 0
+
+        # If size is specified, process in batches until the required size is reached
         if size:
-            aggregation_pipeline.append({"$sample": {"size": size}})
+            while accumulated_size < size:
```
Comment (Senyoret1) [inline orchestration/api/api_external_images.py:361]: I'm not totally sure about what this is trying to do, but it looks like if there are not enough entries accumulated_size will never be equal or more than size, so this will be an infinite loop.

Was this endpoint really working incorrectly and changes were requested?

File: orchestration/api/api_external_images.py [hunk 3]
```diff
@@ -343,25 +342,48 @@ async def get_external_image_list_without_extracts(request: Request, dataset: st
...
+                batch_pipeline = aggregation_pipeline.copy()
+                remaining_size = size - accumulated_size
+                batch_pipeline.append({"$limit": remaining_size})
 
-        # Execute the aggregation
-        image_data_list = list(request.app.external_images_collection.aggregate(aggregation_pipeline))
+                cursor = request.app.external_images_collection.aggregate(batch_pipeline, batchSize=batch_size)
```
Comment (Senyoret1) [inline orchestration/api/api_external_images.py:366]: Is not this excecuting the same query over and over again? I don't see any randomization or pagination, so it looks like this will always return the same elements. For example, if you ask for 5 elements and the first iteration returns [8, 9, 10], the second iteration will return [8, 9] (exactly the same returns, but only 2 elements because the frist iteration already got 3).

File: orchestration/api/api_score.py [hunk 4]
```diff
@@ -22,6 +22,8 @@ async def set_image_rank_score(request: Request, ranking_score: RankingScore):
     if not image_data:
         raise HTTPException(status_code=404, detail="Image with the given hash not found in completed jobs collection.")
 
+    image_uuid = image_data.get('image_uuid', None)
```
Comment (Senyoret1) [inline orchestration/api/api_score.py:25]: I think this won't work, because the query is not retrieving the image_uuid from the database. This would mean saving "null" in the database.

File: orchestration/api/api_score.py [hunk 5]
```diff
@@ -113,6 +116,8 @@ async def set_image_rank_score(
             )
         image_hash = image_data['image_hash']
 
+    image_uuid = image_data.get('image_uuid', None)
```
Comment (Senyoret1) [inline orchestration/api/api_score.py:119]: I think this won't work, because the query is not retrieving the image_uuid from the database. This would mean saving "null" in the database.

File: orchestration/api/api_utils.py [hunk 6]
```diff
@@ -806,40 +821,65 @@ def check_image_usage(request, image_hash):
...
+        # Handle special cases
+        if collection == request.app.all_image_collection:
+            query = {"image_hash": image_hash, "bucket_id": bucket_id}
+            print(f"Removing documents with image_hash: {image_hash} and bucket_id: {bucket_id} from {collection.name}")
+        
+        elif collection in [request.app.image_rank_scores_collection, 
+                            request.app.image_classifier_scores_collection, 
+                            request.app.irrelevant_images_collection]:
+            if collection == request.app.irrelevant_images_collection:
+                query = {"file_hash": image_hash}
```
Comment (Senyoret1) [inline orchestration/api/api_utils.py:861]: "image_source" should be added here too.

File: scripts/update_collection_with_new_uuid.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            if not image_hash:
+                print("Skipping document without image_hash.")
+                continue  # Skip if no image_hash is present
+
+            if "image_uuid" in rank_score:
+                print(f"Skipping document with _id: {rank_score['_id']} as it already has image_uuid.")
+                continue  # Skip if image_uuid is already present
+
+            # Find the corresponding document in completed_jobs_collection
+            job_data = completed_jobs_collection.find_one({"image_hash": image_hash}, {"uuid": 1})
```
Comment (Senyoret1) [inline scripts/update_collection_with_new_uuid.py:32]: Also check the image source.

File: scripts/clean_collections_from_deleted_images.py [hunk 8]
```diff
@@ -2,21 +2,71 @@
...
+def process_collection(collection, minio_client, all_existing_hashes, hash_field, chunk_size=1000):
+    """
+    Process each collection in chunks to avoid DocumentTooLarge errors.
+    """
+    orphaned_count = 0
+
+    # Process in chunks
+    for i in range(0, len(all_existing_hashes), chunk_size):
+        chunk = list(all_existing_hashes)[i:i + chunk_size]
+        orphaned_docs_cursor = collection.find({hash_field: {"$nin": chunk}}).batch_size(2000)
```
Comment (Senyoret1) [inline scripts/clean_collections_from_deleted_images.py:14]: This function will wipe all the data in the collection.

Here the code will get all the data in the collection, including data for images that do exist in "all_existing_hashes". This query will only ignore data from the 1000 hashes in the batch, but everything else will be returned. In the following lines, all that data will be delete in one go. After that, the next iteration of the loop will get the remaining data and will remove all that data too.

So, in summary, after 2 loop iteration the whole collection will end totally empty, with 0 entries.

File: scripts/update_collection_with_new_uuid.py [hunk 9]
```diff
@@ -29,16 +29,15 @@ def add_image_uuid_to_rank_scores():
                 continue  # Skip if image_uuid is already present
 
             # Find the corresponding document in completed_jobs_collection
-            job_data = completed_jobs_collection.find_one({"image_hash": image_hash}, {"uuid": 1})
-            if job_data and "uuid" in job_data:
+            job_data = completed_jobs_collection.find_one({"image_hash": image_hash}, {"uuid": 1, "image_source": 1})
```
Comment (Senyoret1) [inline scripts/update_collection_with_new_uuid.py:32]: This is not what was requested in the previous review. There is no point in getting the "image_source" property, as it is not used in the code. What was requested was to check the image type in the query

File: orchestration/api/api_score.py [hunk 10]
```diff
@@ -14,23 +14,26 @@
...
-        {"task_output_file_dict.output_file_hash": ranking_score.image_hash},
-        {"task_output_file_dict.output_file_hash": 1}
+        {"uuid": ranking_score.uuid},
+        {"image_uuid": 1}
     )
     if not image_data:
-        raise HTTPException(status_code=404, detail="Image with the given hash not found in completed jobs collection.")
+        raise HTTPException(status_code=404, detail="Image with the given uuid not found in completed jobs collection.")
+
+    image_uuid = image_data.get('uuid', None)
```
Comment (Senyoret1) [inline orchestration/api/api_score.py:25]: This is incorrect, the image uuid is in the "image_uuid" field, not the "uuid" field.

File: scripts/update_collection_with_new_uuid.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+            if not image_hash:
+                print("Skipping document without image_hash.")
+                continue  # Skip if no image_hash is present
+
+            if "image_uuid" in rank_score:
+                print(f"Skipping document with _id: {rank_score['_id']} as it already has image_uuid.")
+                continue  # Skip if image_uuid is already present
+
+            # Find the corresponding document in completed_jobs_collection
+            job_data = completed_jobs_collection.find_one({"image_hash": image_hash}, {"uuid": 1, "image_source": 1})
```
Comment (Senyoret1) [inline scripts/update_collection_with_new_uuid.py:32]: This is not what was requested in the previous review. There is no point in getting the "image_source" property, as it is not used in the code. What was requested was to check the image type in the query

File: scripts/clean_collections_from_deleted_images.py [hunk 12]
```diff
@@ -1,82 +1,99 @@
...
-                    for file in associated_files:
-                        minio_client.remove_object(bucket_name, file)
-                        print(f"Removed {file} from {bucket_name}")
-                except Exception as e:
-                    print(f"Failed to remove {object_name} from {bucket_name}: {e}")
-
-        # Remove orphaned documents from MongoDB
+
+        for doc in orphaned_docs_cursor:
+            if collection.name == "all_image_collection":
```
Comment (Senyoret1) [inline scripts/clean_collections_from_deleted_images.py:17]: This will always be calse. The name of the all images collection, in the database, is "all-images", not "all_image_collection"

File: scripts/check_collection_for_different_entries.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
+from pymongo import MongoClient
```
Comment (Senyoret1) [inline scripts/check_collection_for_different_entries.py:1]: The aproach taken by this script is not incorrect, but may be too strict and give false positives. Maybe the script should also count how many entries have the following properties:

- Have a dataset field.
- Have a model_id field.
- Don't have a rank_model_id or rank_id field.

That would help to have a better idea of what kind of invalid data is in the collection, if any.
